sdk/core/System.ClientModel/src/Options/PipelineMessageClassifier.cs

```diff
@@ -1,25 +1,91 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System.IO;
+
 namespace System.ClientModel.Primitives;
 
 public class PipelineMessageClassifier
 {
-    internal static PipelineMessageClassifier Default { get; } = new PipelineMessageClassifier();
+    public static PipelineMessageClassifier Default { get; } = new EndOfChainClassifier();
 
     public static PipelineMessageClassifier Create(ReadOnlySpan<ushort> successStatusCodes)
         => new ResponseStatusClassifier(successStatusCodes);
 
-    protected internal PipelineMessageClassifier() { }
+    protected PipelineMessageClassifier() { }
+
+    public virtual bool TryClassify(PipelineMessage message, out bool isError)
+    {
+        isError = false;
+        return false;
+    }
+
+    public virtual bool TryClassify(PipelineMessage message, Exception? exception, out bool isRetriable)
+    {
+        isRetriable = false;
+        return false;
+    }
 
-    /// <summary>
-    /// Specifies if the response contained in the <paramref name="message"/> is not successful.
-    /// </summary>
-    public virtual bool IsErrorResponse(PipelineMessage message)
+    internal class EndOfChainClassifier : PipelineMessageClassifier
     {
-        message.AssertResponse();
+        public override bool TryClassify(PipelineMessage message, out bool isError)
+        {
+            message.AssertResponse();
+
+            int statusKind = message.Response!.Status / 100;
+            isError = statusKind == 4 || statusKind == 5;
+
+            // Always classify the message
+            return true;
```

Comment:
Why don't we use the bitmap classifier then? It's lower pri, but maybe we should compare perf of the bitmap based and this classifier and if they are close, we could remove this class and just use the other

Reply:
Today, it's expensive to initialize a bitmap classifier with a lot of codes set since set status codes one at a time in a loop.  I think someone closed the outstanding issue I had filed to implement range-sets but I'm happy to open a tracking issue to do the perf comparison and see if that's preferred here.  I don't think it would change the public APIs at all.

Reply:
We would use internal/private APIs to simply use a hardcoded long[]

Reply:
Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41620